kotti/rest.py

```diff
@@ -0,0 +1,306 @@
+from zope.interface import Interface
+import colander
+import datetime
+import decimal
+import json
+import venusian
+
+
+class ISchemaFactory(Interface):
+    """ A factory for colander schemas.
+    """
+
+    def __call__(context, request):
+        """ Returns a colander schema instance """
+
+
+class IContentFactory(Interface):
+    """ A factory for content factories. Can be a class, ex: Document
+    """
+
+    def __call__(context, request):
+        """ Returns a factory that can construct a new object """
+
+
+def _schema_factory_name(context=None, type_name=None, name=u'default'):
+
+    if not any(map(lambda x: x is not None, [context, type_name])):
+        raise Exception("Provide a context or a type name")
+
+    type_name = (context is not None) and context.type_info.name or type_name
+
+    return u"{}/{}".format(type_name, name)
+
+
+def schema_factory(klass, name=u'default'):
+    """ A decorator to be used to mark a function as a serializer.
+
+    The decorated function should return a basic python structure usable (along
+    the lines of colander's cstruct) by a JSON encoder.
+
+    It will also register the context class as a factory for that content.
+    """
+
+    name = _schema_factory_name(context=klass, name=name)
+
+    def wrapper(wrapped):
+        def callback(context, funcname, ob):
+            config = context.config.with_package(info.module)
+            config.registry.registerUtility(wrapped, ISchemaFactory, name=name)
+            config.registry.registerUtility(klass, IContentFactory,
+                                            name=klass.type_info.name)
+
+        info = venusian.attach(wrapped, callback, category='pyramid')
+        return wrapped
+
+    return wrapper
+
+
+@schema_factory(Content)
+def content_schema_factory(context, request):
+    from kotti.views.edit.content import ContentSchema
+    return ContentSchema()
+
+
+@schema_factory(Document)
+def document_schema_factory(context, request):
+    from kotti.views.edit.content import DocumentSchema
+    return DocumentSchema()
+
+
+@schema_factory(File)
+def file_schema_factory(context, request):
+    from kotti.views.edit.content import FileSchema
+    return FileSchema(None)
+
+
+ACCEPT = 'application/vnd.api+json'
+
+
+@view_defaults(name='', accept=ACCEPT, renderer="kotti_jsonp")
+class RestView(object):
+    """ A generic @@json view for any and all contexts.
+
+    Its response depends on the HTTP verb used. For ex:
+    """
+
+    def __init__(self, context, request):
+        self.context = context
+        self.request = request
+
+    @view_config(request_method='GET')
+    def get(self):
+        return self.context
+
+    @view_config(request_method='POST')
+    def post(self):
+        data = self.request.json_body['data']
+
+        assert data['id'] == self.context.name
+        assert data['type'] == self.context.type_info.name
+
+        schema = schema_factory(self.context, name='edit')(
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Too many positional arguments for function call](https://www.codacy.com/app/disko/Kotti/file/988961580/issues/source?bid=2089747&fileBranchId=2861988#l141)